server/src/main/java/org/elasticsearch/index/mapper/StoredValueFetcher.java

```diff
@@ -38,7 +38,12 @@ public void setNextReader(LeafReaderContext context) {
     @Override
     public List<Object> fetchValues(Source source, int doc, List<Object> ignoredValues) throws IOException {
         leafSearchLookup.setDocument(doc);
-        return leafSearchLookup.fields().get(fieldname).getValues();
+        List<Object> values = leafSearchLookup.fields().get(fieldname).getValues();
+        if (IdFieldMapper.NAME.equals(fieldname)) {
```

Comment:
I don't think we need to be that specific. StoredValueFetcher is only used in the IgnoredFieldMapper which probably also suffers from the cache clearing issue (maybe makes sense to add a test?), the LegacyTypeFieldMapper (from ES5) and otherwise only the IdFieldMapper and subtypes. Let's just return a List.copyOf here in all cases.
```suggestion
        return List.copyOf(values);
```